### PR TITLE
[P8-106] Add DependencyHeuristicSearch

### PR DESCRIPTION
**Pull request by NicEastvillage**
A heuristic inspired by PageRank. Vertices with more dependencies are more important, hence edges of this vertex is more important.

### CODE REVIEW

**Review by GeneralDenmark — APPROVED** [review]
With the exception of newly introduced Clippy warnings, it looks good to me :+1:

File: atl-checker/src/algorithms/certain_zero/search_strategy/dependency_heuristic.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        for edge in edges {
+            // For each target, increase their priority since we found a new edge,
+            // that depends on it
+            match &edge {
+                Edge::HYPER(edge) => {
+                    for target in &edge.targets {
+                        let dependency = self
+                            .dependencies
+                            .entry(target.clone())
+                            .or_insert(DependencyData::new());
```
Comment (lars4509) [inline atl-checker/src/algorithms/certain_zero/search_strategy/dependency_heuristic.rs:82]: I am pretty sure, you can fix the warning just by using `or_insert_with()` as suggested.

File: atl-checker/src/algorithms/certain_zero/search_strategy/dependency_heuristic.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                        dependency.rank += 1;
+                    }
+                }
+                Edge::NEGATION(edge) => {
+                    let dependency = self
+                        .dependencies
+                        .entry(edge.target.clone())
+                        .or_insert(DependencyData::new());
```
Comment (lars4509) [inline atl-checker/src/algorithms/certain_zero/search_strategy/dependency_heuristic.rs:90]: Same here

File: atl-checker/src/algorithms/certain_zero/search_strategy/dependency_heuristic.rs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    fn queue_back_propagation(&mut self, edge: Edge<V>) {
+        self.queue(edge);
+    }
+
+    fn on_interest(&mut self, vertex: &V) {
+        let dependency = self
+            .dependencies
+            .entry(vertex.clone())
+            .or_insert(DependencyData::new());
```
Comment (lars4509) [inline atl-checker/src/algorithms/certain_zero/search_strategy/dependency_heuristic.rs:113]: and here

**Review by lars4509 — APPROVED** [review]
Looking away from the clippy it looks good to me.

**Comment by NicEastvillage** [comment]
The clippy man strikes again!

**Review by MathiasMehl — APPROVED** [review]

**Comment by falkecarlsen** [comment]
@NicEastvillage needs merge-conflict fix please

**Pull request merged**